### PR TITLE
SDI-554 create sentence adjustment mapping error handling

### PR DESCRIPTION
**Pull request by andymarke**
* Since we send our selves a retry message we do that to the "migration" queue - this has lead us to hack the MigrationContext so this PR corrects that with a new SynchronisationContext to pass data around that is not part of the body of the message

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/data/MessageContexts.kt [hunk 1]
```diff
@@ -14,3 +19,9 @@ class MigrationContext<T>(val type: MigrationType, val migrationId: String, val
 }
 
 fun generateBatchId(): String = LocalDateTime.now().withNano(0).format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)
+
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/data/MessageContexts.kt:22]: First use is when sentencing synchronisation posts a message to itself to retry mapping, we no longer have to add dummy "migration" attributes

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/incentives/MigrationIncentivesMessageListener.kt [hunk 2]
```diff
@@ -20,7 +20,7 @@ import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.incentives.Incent
 import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.incentives.IncentiveMessages.RETRY_INCENTIVE_MAPPING
 import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.incentives.IncentiveMessages.RETRY_INCENTIVE_SYNCHRONISATION_MAPPING
 import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.incentives.IncentiveMessages.SYNCHRONISE_CURRENT_INCENTIVE
-import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.listeners.context
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.listeners.migrationContext
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/incentives/MigrationIncentivesMessageListener.kt:23]: just a rename now with 2 types on context

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/sentencing/MigrationSentencingMessageListener.kt [hunk 3]
```diff
@@ -37,28 +40,31 @@ class MigrationSentencingMessageListener(
   @WithSpan(value = "dps-syscon-migration_sentencing_queue", kind = SpanKind.SERVER)
   fun onMessage(message: String, rawMessage: Message): CompletableFuture<Void>? {
     log.debug("Received message {}", message)
-    val migrationMessage: MigrationMessage<SentencingMessages, *> = message.fromJson()
+    val migrationMessage: LocalMessage<SentencingMessages> = message.fromJson()
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/sentencing/MigrationSentencingMessageListener.kt:43]: base class for both types on message types, Migration and Synchronisation

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/sentencing/MigrationSentencingMessageListener.kt [hunk 4]
```diff
@@ -37,28 +40,31 @@ class MigrationSentencingMessageListener(
...
+          MIGRATE_SENTENCING_ADJUSTMENT -> sentencingMigrationService.migrateSentencingAdjustment(migrationContext(message.fromJson()))
+          MIGRATE_SENTENCING_STATUS_CHECK -> sentencingMigrationService.migrateSentencingStatusCheck(migrationContext(message.fromJson()))
+          CANCEL_MIGRATE_SENTENCING -> sentencingMigrationService.cancelMigrateSentencingStatusCheck(migrationContext(message.fromJson()))
+          RETRY_MIGRATION_SENTENCING_ADJUSTMENT_MAPPING -> sentencingMigrationService.retryCreateSentenceAdjustmentMapping(
+            migrationContext(
+              message.fromJson()
+            )
+          )
+
+          RETRY_SYNCHRONISATION_SENTENCING_ADJUSTMENT_MAPPING -> sentencingSynchronisationService.retryCreateSentenceAdjustmentMapping(
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/sentencing/MigrationSentencingMessageListener.kt:63]: actual new code

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/sentencing/SentencingMessages.kt [hunk 5]
```diff
@@ -6,5 +6,6 @@ enum class SentencingMessages {
   MIGRATE_SENTENCING_ADJUSTMENT,
   MIGRATE_SENTENCING_STATUS_CHECK, // status check and cancel work at queue level. The queue is used by all Sentencing migration entity migrations
   CANCEL_MIGRATE_SENTENCING,
-  RETRY_SENTENCING_ADJUSTMENT_MAPPING,
+  RETRY_MIGRATION_SENTENCING_ADJUSTMENT_MAPPING,
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/sentencing/SentencingMessages.kt:9]: rename to distinguish from the sync similar failure

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/sentencing/SentencingSynchronisationService.kt [hunk 6]
```diff
@@ -31,25 +43,77 @@ class SentencingSynchronisationService(
...
       }
     }
   }
+
+  enum class MappingResponse {
+    MAPPING_CREATED,
+    MAPPING_FAILED
+  }
+
+  suspend fun tryToCreateMapping(
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/sentencing/SentencingSynchronisationService.kt:62]: main change here, send ourselves a message if it fails so it eventually succeeds

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/service/Helpers.kt [hunk 7]
```diff
@@ -1,22 +1,33 @@
...
 import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.data.MigrationContext
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.data.SynchronisationContext
 import java.time.LocalDate
 import java.time.LocalDateTime
 import java.time.format.DateTimeFormatter
 
 fun LocalDateTime?.asStringOrBlank(): String = this?.format(DateTimeFormatter.ISO_DATE_TIME) ?: ""
 fun LocalDate?.asStringOrBlank(): String = this?.format(DateTimeFormatter.ISO_DATE) ?: ""
 
-class MigrationMessage<M, T>(
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/service/Helpers.kt:11]: should have called this Message but it clashed with too many other classes called Message to panicked and called it this instead

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/service/Helpers.kt [hunk 8]
```diff
@@ -1,22 +1,33 @@
...
+  val context: SynchronisationContext<T>
+) : LocalMessage<M>(type)
+
 const val VISITS_QUEUE_ID = "migrationvisits"
 const val INCENTIVES_QUEUE_ID = "migrationincentives"
 const val SENTENCING_QUEUE_ID = "migrationsentencing"
-enum class MigrationType(val queueId: String) {
+
+enum class SynchronisationType(val queueId: String) {
   VISITS(VISITS_QUEUE_ID),
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/service/Helpers.kt:31]: This word can be used fort batch Migration and one by one Synchronisation

**Review by steverendell — APPROVED** [review]
lgtm

**Pull request merged**